### ISSUE TITLE
Any training results available?

### ISSUE THREAD

**Issue by ruiyuanlu** [cloneofsimo/smallest_working_performer#1]
Hi, thanks for sharing. Are there any training results (Cifar, MNIST etc.) available?

**Comment by cloneofsimo** [comment]
Roughly with the settings you should get something like 60% accuracy on CIFAR - 10, within 20 epochs or so. You also get 90% accuracy on MNIST within 3 epochs. Greater kernel_ratio yields more accurate exp(Q^tK), but requires more time. Transformers requires much more architecture searching and hyperparameter tuning, so I'm 100% sure you can get better results with ViTs, even with this performer model.
Partly this is because ViTs, as authors noticed, takes long time to train than other CNN models, because CNNs can quickly exploit trivial - local patterns.

**Comment by lucastononrodrigues** [comment]
Did you train it yourself? I mean using this code?
By pretty much using the same logic and readapting the code (not changing any fundamental operation) I get nans after the third layer. I will check if I'm missing anything.

EDIT: by normalizing the data before the exponencial function by multiplying by q.shape[-1]**(-0.25) I solved the problem.

**Comment by cloneofsimo** [comment]
Adding note,
Using the exact code from my source code, with settings
```python
model = ViP(
    image_pix = 28,
    patch_pix = 4, # this will result in 14 * 14 words
    class_cnt = 10,
    layer_cnt = 3,
    kernel_ratio = 2
)
device = "cuda:0"
batch_size = 128
lr = 2e-4
epochs = 3
```

```python
emb_s = 32
head_cnt = 6
dp1 = 0.1
dp2 = 0.1
dp3 = 0.1
```

with seed 0, epoch 3, should give you average loss : 0.1967 with test_acc : 0.9516 (95%)
Also It should take less than 3 min for the whole process on modern gpus.
In the paper, they used something like 256 features if I remember correctly. Having kernel_ratio = 2 gives you good approximating kernel from my experiments
